zebra-chain/src/value_balance.rs

```diff
@@ -333,55 +319,80 @@ impl ValueBalance<NonNegative> {
     }
 
     /// To byte array
-    pub fn to_bytes(self) -> [u8; 32] {
-        let transparent = self.transparent.to_bytes();
-        let sprout = self.sprout.to_bytes();
-        let sapling = self.sapling.to_bytes();
-        let orchard = self.orchard.to_bytes();
-        match [transparent, sprout, sapling, orchard].concat().try_into() {
+    pub fn to_bytes(self) -> [u8; 40] {
+        match [
+            self.transparent.to_bytes(),
+            self.sprout.to_bytes(),
+            self.sapling.to_bytes(),
+            self.orchard.to_bytes(),
+            self.deferred.to_bytes(),
+        ]
+        .concat()
+        .try_into()
+        {
             Ok(bytes) => bytes,
             _ => unreachable!(
-                "Four [u8; 8] should always concat with no error into a single [u8; 32]"
+                "five [u8; 8] should always concat with no error into a single [u8; 40]"
             ),
         }
     }
 
     /// From byte array
     #[allow(clippy::unwrap_in_result)]
     pub fn from_bytes(bytes: &[u8]) -> Result<ValueBalance<NonNegative>, ValueBalanceError> {
+        let bytes_length = bytes.len();
+
+        // Return an error early if bytes don't have the right lenght instead of panicking later.
+        match bytes_length {
+            32 | 40 => {}
+            _ => return Err(Unparsable),
+        };
```

Comment:
###### Optional / Nitpick

We could combine the match statements in this method if we're okay with assigning bytes out of order:

```rust
let deferred = match bytes.len() {
    32 => Amount::zero(),
    40 => Amount::from_bytes(
        bytes[32..40]
            .try_into()
            .expect("deferred amount should be parsable"),
    )
    .map_err(Deferred)?,
    _ => return Err(Unparsable),
};
```

Reply:
I prefer having the check first and keeping reading the bytes in order.